anydi/_container.py

```diff
@@ -269,151 +165,89 @@ def _get_or_register_provider(self, interface: AnyInterface) -> Provider:
-                "The scope provided is invalid. Only the following scopes are "
-                f"supported: {', '.join(get_args(Scope))}. Please use one of the "
-                "supported scopes when registering a provider."
-            )
-
-    def _validate_provider_type(self, provider: Provider) -> None:
-        """Validate the type of provider.
-
-        Args:
-            provider: The provider to validate.
-
-        Raises:
-            TypeError: If the provider has an invalid type.
-        """
-        if provider.is_function or provider.is_class:
-            return
+            self._resource_cache[provider.scope].append(provider.interface)
 
+    def _delete_provider(self, provider: Provider) -> None:
+        """Delete a provider."""
+        if provider.interface in self._providers:
+            del self._providers[provider.interface]
         if provider.is_resource:
-            if provider.scope == "transient":
-                raise TypeError(
-                    f"The resource provider `{provider}` is attempting to register "
-                    "with a transient scope, which is not allowed. Please update the "
-                    "provider's scope to an appropriate value before registering it."
-                )
-            return
+            self._resource_cache[provider.scope].remove(provider.interface)
 
-        raise TypeError(
-            f"The provider `{provider.obj}` is invalid because it is not a callable "
-            "object. Only callable providers are allowed. Please update the provider "
-            "to a callable object before attempting to register it."
-        )
-
-    def _validate_provider_match_scopes(self, provider: Provider) -> None:
-        """Validate that the provider and its dependencies have matching scopes.
-
-        Args:
-            provider: The provider to validate.
-
-        Raises:
-            ValueError: If the provider and its dependencies have mismatched scopes.
-            TypeError: If a dependency is missing an annotation.
-        """
+    def _validate_sub_providers(self, provider: Provider) -> None:
+        """Validate the sub-providers of a provider."""
 
         for parameter in provider.parameters:
-            if parameter.annotation is inspect.Parameter.empty:
+            annotation = parameter.annotation
+
+            if annotation is inspect.Parameter.empty:
                 raise TypeError(
                     f"Missing provider `{provider}` "
                     f"dependency `{parameter.name}` annotation."
                 )
 
             try:
-                sub_provider = self._get_or_register_provider(parameter.annotation)
+                sub_provider = self._get_or_register_provider(
+                    annotation, parent_scope=provider.scope
+                )
             except LookupError:
+                if provider.scope not in {"singleton", "transient"}:
+                    self._unresolved_interfaces.add(provider.interface)
```

Comment:
If I get it correctly, now the delayed registration of `request` scoped stuff is possible (mainly intended for instances, but affects real providers too).

This seems to be the only modification of this set. Shouldn't there be an `_unresolved_interfaces.remove` at some point?
For example, if this is a `request` scoped provider but it is registered later in time, we should remove it from this set, or?

Reply:
Alternatively, you could consider requiring the pre-registration of placeholders for request scoped instances.   
For example `container.register(MyRequestScopedType, anydi.somepackage.RequestScopedInstancePlaceholder())` or even `container.declare_request_scoped_variable(MyRequestScopedType)`.  
In this case no lenient validation is needed anymore for request scope.